In [1]:
import numpy as np
import pandas as pd

# Citation Racial Breakdown

A pretty straight forward process where I took the PDFs given of the citations in 2019 and 2020 (given in 2 seperate documents), converted to a csv file using https://convertio.co/pdf-csv/, and then grouped by race or sex to count the number for each category.

In [2]:
file = "fwdfwdatatransparency/2019 Citation Data.csv"
citations19 = pd.read_csv(file)

#set first row's labels to be column names
citations19.columns = ['Report Number', 'Date', 'Violation Code', 'Race', 'Sex']
citations19 = citations19.iloc[1:]

#fill in empty cells
citations19['Race'] = citations19['Race'].fillna("Not Recorded")

citations19

Report Number        Date        Violation Code          Race   Sex
1          T503714    1/1/2019                 24601         Black  Male
2          T506100    1/1/2019        267085A 12500A      Hispanic  Male
3          E219189    1/2/2019         23123A 16028A       Unknown     M
4          E219190    1/2/2019                21453A       Unknown     F
5          E219191    1/2/2019                21453A       Unknown     M
...            ...         ...                   ...           ...   ...
9516       E230705  12/31/2019  23109B 16028A 40004A         Other     M
9517       E233348  12/31/2019                21453A   Other Asian     M
9518       E233729  12/31/2019         4000A1 23109B  Not Recorded     F
9519       E234438  12/31/2019                12500A      Hispanic     M
9520       T508895  12/31/2019                4000A1         White  Male

[9520 rows x 5 columns]

In [3]:
#group by race, count number per race category
race19 = citations19.groupby(['Race']).count()

#reset index
race19.reset_index(inplace=True)

#leave only Race and Count as the 2 columns
race19 = race19[['Race', 'Report Number']]
race19.columns = ['Race', 'Count']
race19.name = "Citations By Race in 2019"

race19 = pd.DataFrame(race19)
race19

Race  Count
0    American Indian     11
1       Asian Indian    888
2              Black    229
3            Chinese      7
4           Filipino      6
5           Hispanic   1561
6       Not Recorded   2747
7              Other    681
8        Other Asian   1302
9   Pacific Islander      2
10           Unknown    680
11        Vietnamese      3
12             White   1403

In [4]:
file = "fwdfwdatatransparency/2020 Citation Data.csv"
citations20 = pd.read_csv(file)

#set first row's labels to be column names
citations20.columns = ['Report Number', 'Date', 'Violation Code', 'Race', 'Sex']
citations20 = citations20.iloc[1:]

##fill in empty cells 
citations20['Race'] = citations20['Race'].fillna("Not Recorded")

citations20

Report Number       Date        Violation Code          Race   Sex
1          T510608   1/1/2020                24252A      Hispanic  Male
2          E234202   1/2/2020                12500A         Other     M
3          E235931   1/2/2020               231235A  Not Recorded     M
4          E235932   1/2/2020                24252A  Not Recorded     F
5          E235933   1/2/2020                22450A  Not Recorded     F
...            ...        ...                   ...           ...   ...
1339       E234645  6/28/2020                21453A  Not Recorded     M
1340       E234646  6/28/2020  12500A 4000A1 16028A      Hispanic     M
1341       E234647  6/28/2020                21453A         White     M
1342       E234648  6/29/2020          22350 12500A      Hispanic     M
1343       E234479  6/30/2020                27151A   Other Asian     M

[1343 rows x 5 columns]

In [5]:
#group by race, count number per race category
race20 = citations20.groupby(['Race']).count()

#reset index
race20.reset_index(inplace=True)

#leave only Race and Count as the 2 columns
race20 = race20[['Race', 'Report Number']]
race20.columns = ['Race', 'Count']
race20.name = "Citations By Race in 2020"

race20 = pd.DataFrame(race20)
race20

Race  Count
0    American Indian      3
1       Asian Indian     99
2              Black     38
3           Filipino      1
4           Hispanic    226
5       Not Recorded    420
6              Other     83
7        Other Asian    175
8   Pacific Islander      1
9            Unknown    109
10        Vietnamese      1
11             White    187

In [6]:
#combine 2019 and 2020 tables into one large table
citations1920 = pd.concat([citations19, citations20])

In [7]:
#group by race, count number per race category
citations_race1920 = citations1920.groupby(['Race']).count()

#reset index
citations_race1920.reset_index(inplace=True)

#leave only Race and Count as the 2 columns
citations_race1920 = citations_race1920[['Race', 'Report Number']]
citations_race1920.columns = ['Race', 'Count']
citations_race1920.title = "Citations By Race in 2019 and 2020 Combined"

citations_race1920 = pd.DataFrame(citations_race1920)
citations_race1920

Race  Count
0    American Indian     14
1       Asian Indian    987
2              Black    267
3            Chinese      7
4           Filipino      7
5           Hispanic   1787
6       Not Recorded   3167
7              Other    764
8        Other Asian   1477
9   Pacific Islander      3
10           Unknown    789
11        Vietnamese      4
12             White   1590

# Arrest Race Breakdown

Similar process as what was explained for the citations breakdown, the only difference is that the data was more broken up so I had to concat all of the tables. Chose not to do a seperate 2019 and 2020 table, but let me know if it would be helpful. 

In [8]:
file_names = ["fwdfwdatatransparency/convert/01012019-01312019-ARREST-DATA_Redacted.csv",
              "fwdfwdatatransparency/convert/01012020-01312020-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/02012019-02282019-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/02012020-02292020-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/03012019-05312019-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/03012020-05312020-ARREST-DATA_Redacted.csv",
              "fwdfwdatatransparency/convert/06012019-08312019-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/06012020-06302020-ARREST-DATA_Redacted.csv", 
              "fwdfwdatatransparency/convert/09012019-10312019-ARREST-DATA_Redacted.csv",
              "fwdfwdatatransparency/convert/09012019-10312019-ARREST-DATA_Redacted.csv",
              "fwdfwdatatransparency/convert/11012019-12312019-ARREST-DATA_Redacted.csv" ]
arrests = []
for file in file_names:
    arrest = pd.read_csv(file)
    
    #set column name
    arrest.columns = ['Race']
    
    #drop empty rows (because of how converting from pdf to csv resulted in a lot of empty rows)
    arrest = arrest.dropna()
    
    #drop rows w not a race label 
    arrest = arrest[arrest['Race'].map(len) < 2]
    
    #remove last empty row
    arrest = arrest[:-1]
    
    arrests.append(arrest)

In [9]:
#combine all sub arrest tables into one table
arrests1920 = pd.concat(arrests)

#change race labels based off table given by Sunnyvale PD
arrests1920['Race'].replace(['A', 'B', 'C', 'F', 'G', 'H', 'I', 'K', 'O', 'P', 'S', 'V', 'W', 'X', 'Z'], ['Other Asian', 'Black', 'Chinese', 'Filipino', 'Guamanian', 'Hispanic/Mexican/Latin', 'American Indian/Alaskan', 'Korean', 'Other', 'Native Hawaiian/Other Pacific', 'Samoan', 'Vietnamese', 'White', 'Unknown', 'Asian Indian'], inplace = True)

arrests1920 = arrests1920.dropna()

In [10]:
#group by race, count number per race category
arrests = arrests1920.groupby(['Race'])['Race'].count()
arrests = pd.DataFrame(arrests)
arrests.columns = ['Count']
arrests.reset_index(inplace = True)
arrests

Race  Count
0         American Indian/Alaskan      2
1                    Asian Indian    157
2                           Black    506
3                         Chinese     11
4                        Filipino     21
5                       Guamanian      3
6          Hispanic/Mexican/Latin   1803
7                          Korean      1
8   Native Hawaiian/Other Pacific     13
9                           Other     89
10                    Other Asian    219
11                         Samoan      7
12                        Unknown     13
13                     Vietnamese      7
14                          White   1144

# Approximation of Likelihood To Be Arrested/Cited Based on Race

Based on the Los Altos released data breakdown of the ratio of arrests/citations of arrests vs. nonresidents, this will use the same ratio to approximate the breakdown of residents vs. non residents in the Sunnyvale data. This is because otherwise, there is no way to approximate the likelihood of being arrested/cited on the basis of race. 

Additionally, it is a non-trivial issue to approximate the likelihood of being arrested/cited because there is no way of knowing the exact demographics of those who pass through Sunnyvale. Thus, to bes estimate, I am relying on Los Altos police data, which shows a breakdown of resident vs. non resident. For each category of race, I am solving for the equation (# of arrests or citations of resident) / (# of arrests or citations of non-resident) = (# of residents) / (# non residents) and solving for the # of non residents to get an approximation of the number of non residents of each race that pass through the city. 

I also chose to do non resident estimates by race because of the likely possibility that different races have different traveling rates (TODO: insert proof for statistical significance here)

Note: This all relies on the assumption that, given your race, the probability of being arrested does not depend on your residential status. (TODO: insert proof for why that is here)

Obviously, this is not exact, and it would be helpful to get Sunnyvale data on the exact proportion of arrests vs. non arrests.

My initial instinct was to try to use the demographics of the cities directly neighboring Sunnyvale to estimate the demographics of those that would pass through Sunnyvale. However, I think the above method is a more accurate assumption than to rely on the demographics of neighboring cities. (I'd started on the code for this process, so I am going to keep it for reference at the botto mof the page.

In [11]:
sunnyvale_pop = 152703
sunnyvale = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [sunnyvale_pop*0.312, sunnyvale_pop*0.017, sunnyvale_pop*0.004, sunnyvale_pop*0.459, sunnyvale_pop*0.002, sunnyvale_pop*0.044, sunnyvale_pop*0.173]})
sunnyvale

Race  Number of Residents
0                    White            47643.336
1                    Black             2595.951
2  American Indian/Alaskan              610.812
3                    Asian            70090.677
4          Native Hawaiian              305.406
5              Two or More             6718.932
6                 Hispanic            26417.619

In [12]:
print("Sunnyvale July 1, 2019 Population Estimate: " + str(sunnyvale_pop))
print("Total sum of calculated number of residents: "+ str(sum(sunnyvale['Number of Residents'])))

Sunnyvale July 1, 2019 Population Estimate: 152703
Total sum of calculated number of residents: 154382.733


## Using Los Altos data to get number of non residents by race

In [13]:
los_altos_pop = 30089
los_altos = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Count': [los_altos_pop*0.605, los_altos_pop*0.004, los_altos_pop*0.002, los_altos_pop*0.294, los_altos_pop*0, los_altos_pop*0.052, los_altos_pop*0.043]})
los_altos

Race      Count
0                    White  18203.845
1                    Black    120.356
2  American Indian/Alaskan     60.178
3                    Asian   8846.166
4          Native Hawaiian      0.000
5              Two or More   1564.628
6                 Hispanic   1293.827

In [14]:
print("Los Altos July 1, 2019 Population Estimate: " + str(los_altos_pop))
print("Total sum of calculated number of residents: "+ str(sum(los_altos['Count'])))

Los Altos July 1, 2019 Population Estimate: 30089
Total sum of calculated number of residents: 30089.0


### Los Altos Arrest Data

In [15]:
los_altos_resident_arrests = pd.DataFrame(
    { 'Race' : ['Asian', 'Black', 'Chinese', 'Guamanian', 'Hispanic', 'Korean', 'Other', 'Hawaiian', 'Vietnamese', 'White', 'Unknown'],
      'Count' : [16, 4, 1, 0, 12, 0, 12, 0, 0, 65, 0]
    })
los_altos_resident_arrests

Race  Count
0        Asian     16
1        Black      4
2      Chinese      1
3    Guamanian      0
4     Hispanic     12
5       Korean      0
6        Other     12
7     Hawaiian      0
8   Vietnamese      0
9        White     65
10     Unknown      0

In [16]:
los_altos_nonresident_arrests = pd.DataFrame(
    { 'Race' : ['Asian', 'Black', 'Chinese', 'Guamanian', 'Hispanic', 'Korean', 'Other', 'Hawaiian', 'Vietnamese', 'White', 'Unknown'],
      'Count' : [26, 44, 0, 1, 189, 1, 44, 1, 1, 155, 2]
    })
los_altos_nonresident_arrests

Race  Count
0        Asian     26
1        Black     44
2      Chinese      0
3    Guamanian      1
4     Hispanic    189
5       Korean      1
6        Other     44
7     Hawaiian      1
8   Vietnamese      1
9        White    155
10     Unknown      2

In [17]:
#sanity check to make sure adds up to reported number of arrests
sum(los_altos_resident_arrests['Count']) + sum(los_altos_nonresident_arrests['Count'])

574

In [18]:
los_altos_resident_arrests

Race  Count
0        Asian     16
1        Black      4
2      Chinese      1
3    Guamanian      0
4     Hispanic     12
5       Korean      0
6        Other     12
7     Hawaiian      0
8   Vietnamese      0
9        White     65
10     Unknown      0

In [19]:
census_races_of_los_altos_resident_arrest = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(los_altos_resident_arrests[los_altos_resident_arrests['Race'] == 'White']['Count']), 
                             int(los_altos_resident_arrests[los_altos_resident_arrests['Race'] == 'Black']['Count']), 
                             0, 
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Asian', 'Chinese', 'Korean', 'Vietnamese'])]['Count'].sum()),
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Guamanian', 'Hawaiian'])]['Count'].sum()),  
                             0, 
                             int(los_altos_resident_arrests[los_altos_resident_arrests['Race'] == 'Hispanic']['Count'])]})
census_races_of_los_altos_resident_arrest

Race  Number of Residents
0                    White                   65
1                    Black                    4
2  American Indian/Alaskan                    0
3                    Asian                   28
4          Native Hawaiian                    2
5              Two or More                    0
6                 Hispanic                   12

In [20]:
census_races_of_los_altos_nonresident_arrest = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'] == 'White']['Count']), 
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'] == 'Black']['Count']), 
                             0, 
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Asian', 'Chinese', 'Korean', 'Vietnamese'])]['Count'].sum()),
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Guamanian', 'Hawaiian'])]['Count'].sum()), 
                             0, 
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'] == 'Hispanic']['Count'])]})
census_races_of_los_altos_nonresident_arrest

Race  Number of Residents
0                    White                  155
1                    Black                   44
2  American Indian/Alaskan                    0
3                    Asian                   28
4          Native Hawaiian                    2
5              Two or More                    0
6                 Hispanic                  189

### Los Altos Citation Data

In [21]:
los_altos_resident_citations = pd.DataFrame(
    { 'Race' : ['Asian', 'Black', 'Chinese', 'Filipino', 'Guamanian', 'Hispanic', 'American Indian', 'Other', 'Pacific Islander', 'Samoan', 'Hawaiian', 'White', 'Unknown', 'Asian Indian'],
      'Count' : [299, 16, 2, 1, 0, 47, 2, 303, 0, 0, 6, 603, 1, 2]
    })
los_altos_resident_citations

Race  Count
0              Asian    299
1              Black     16
2            Chinese      2
3           Filipino      1
4          Guamanian      0
5           Hispanic     47
6    American Indian      2
7              Other    303
8   Pacific Islander      0
9             Samoan      0
10          Hawaiian      6
11             White    603
12           Unknown      1
13      Asian Indian      2

In [22]:
los_altos_nonresident_citations = pd.DataFrame(
    { 'Race' : ['Asian', 'Black', 'Chinese', 'Filipino', 'Guamanian', 'Hispanic', 'American Indian', 'Other', 'Pacific Islander', 'Samoan', 'Hawaiian', 'White', 'Unknown', 'Asian Indian'],
      'Count' : [968, 134, 3, 1, 3, 1337, 3, 1023, 2, 1, 22, 1797, 16, 5]
    })
los_altos_nonresident_citations

Race  Count
0              Asian    968
1              Black    134
2            Chinese      3
3           Filipino      1
4          Guamanian      3
5           Hispanic   1337
6    American Indian      3
7              Other   1023
8   Pacific Islander      2
9             Samoan      1
10          Hawaiian     22
11             White   1797
12           Unknown     16
13      Asian Indian      5

In [23]:
#sanity check to make sure adds up to reported number of citations
sum(los_altos_resident_citations['Count']) + sum(los_altos_nonresident_citations['Count'])

6597

In [24]:
census_races_of_los_altos_resident_citation = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(los_altos_resident_citations[los_altos_resident_citations['Race'] == 'White']['Count']), 
                             int(los_altos_resident_citations[los_altos_resident_citations['Race'] == 'Black']['Count']), 
                             int(los_altos_resident_citations[los_altos_resident_citations['Race'] == 'American Indian']['Count']), 
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Asian', 'Chinese', 'Filipino', 'Asian Indian'])]['Count'].sum()),
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Guamanian', 'Pacific Islander', 'Samoan', 'Hawaiian'])]['Count'].sum()), 
                             0, 
                             int(los_altos_resident_citations[los_altos_resident_citations['Race'] == 'Hispanic']['Count'])]})
census_races_of_los_altos_resident_citation

Race  Number of Residents
0                    White                  603
1                    Black                   16
2  American Indian/Alaskan                    2
3                    Asian                   26
4          Native Hawaiian                    2
5              Two or More                    0
6                 Hispanic                   47

In [25]:
census_races_of_los_altos_nonresident_citation = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(los_altos_nonresident_citations[los_altos_nonresident_citations['Race'] == 'White']['Count']), 
                             int(los_altos_nonresident_citations[los_altos_nonresident_citations['Race'] == 'Black']['Count']), 
                             int(los_altos_nonresident_citations[los_altos_nonresident_citations['Race'] == 'American Indian']['Count']), 
                              int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Asian', 'Chinese', 'Filipino', 'Asian Indian'])]['Count'].sum()),
                             int(los_altos_nonresident_arrests[los_altos_nonresident_arrests['Race'].isin(['Guamanian', 'Pacific Islander', 'Samoan', 'Hawaiian'])]['Count'].sum()), 
                             0, 
                             int(los_altos_nonresident_citations[los_altos_nonresident_citations['Race'] == 'Hispanic']['Count'])]})
census_races_of_los_altos_nonresident_citation

Race  Number of Residents
0                    White                 1797
1                    Black                  134
2  American Indian/Alaskan                    3
3                    Asian                   26
4          Native Hawaiian                    2
5              Two or More                    0
6                 Hispanic                 1337

### Combine arrest and citation data to get number of residents or non residents
I chose to combine citation and arrest data to get a larger sample size to result in a more accurate prediction of the rate of non residents. This relies on the assumption that non residents and residents are arrested and cited with equal probability. 

In [26]:
races_of_citation_and_arrest_resident = census_races_of_los_altos_resident_citation.merge(census_races_of_los_altos_resident_arrest, on = "Race")
races_of_citation_and_arrest_resident['Number of Residents'] = races_of_citation_and_arrest_resident['Number of Residents_x'] + races_of_citation_and_arrest_resident['Number of Residents_y']
races_of_citation_and_arrest_resident = races_of_citation_and_arrest_resident[['Race', 'Number of Residents']]
races_of_citation_and_arrest_resident

Race  Number of Residents
0                    White                  668
1                    Black                   20
2  American Indian/Alaskan                    2
3                    Asian                   54
4          Native Hawaiian                    4
5              Two or More                    0
6                 Hispanic                   59

In [27]:
races_of_citation_and_arrest_nonresident = census_races_of_los_altos_nonresident_citation.merge(census_races_of_los_altos_nonresident_arrest, on = "Race")
races_of_citation_and_arrest_nonresident['Number of Residents'] = races_of_citation_and_arrest_nonresident['Number of Residents_x'] + races_of_citation_and_arrest_nonresident['Number of Residents_y']
races_of_citation_and_arrest_nonresident = races_of_citation_and_arrest_nonresident[['Race', 'Number of Residents']]
races_of_citation_and_arrest_nonresident

Race  Number of Residents
0                    White                 1952
1                    Black                  178
2  American Indian/Alaskan                    3
3                    Asian                   54
4          Native Hawaiian                    4
5              Two or More                    0
6                 Hispanic                 1526

I then divided the number of non residents by the number of residents to get the rate of non residents per rate

In [28]:
nonres_rate = races_of_citation_and_arrest_nonresident.merge(races_of_citation_and_arrest_resident, on = "Race")
nonres_rate['Non Res / Res'] = nonres_rate['Number of Residents_x']/nonres_rate['Number of Residents_y']
nonres_rate = nonres_rate[['Race', 'Non Res / Res']]
nonres_rate

Race  Non Res / Res
0                    White       2.922156
1                    Black       8.900000
2  American Indian/Alaskan       1.500000
3                    Asian       1.000000
4          Native Hawaiian       1.000000
5              Two or More            NaN
6                 Hispanic      25.864407

Given the rate of non residents, I estimated the number of Sunnyvale non residents by multiplying the rate with the sunnyvale population

In [29]:
#calculate estimated sunnyvale number of non residents per race
non_sunnyvale_pop = sunnyvale.merge(nonres_rate, on = "Race")
non_sunnyvale_pop['Non Residents'] = non_sunnyvale_pop['Number of Residents'] * non_sunnyvale_pop['Non Res / Res']
non_sunnyvale_pop = non_sunnyvale_pop[['Race', 'Non Residents']]
non_sunnyvale_pop

Race  Non Residents
0                    White  139221.245317
1                    Black   23103.963900
2  American Indian/Alaskan     916.218000
3                    Asian   70090.677000
4          Native Hawaiian     305.406000
5              Two or More            NaN
6                 Hispanic  683276.043966

I then added the calculated number of non residents to the number of residents to get the total estimated size of each population

In [30]:
sunnyvale_total_pop = non_sunnyvale_pop.merge(sunnyvale, on = "Race")
sunnyvale_total_pop['Total People'] = sunnyvale_total_pop['Non Residents'] + sunnyvale_total_pop['Number of Residents']
sunnyvale_total_pop = sunnyvale_total_pop[['Race', 'Total People']]
sunnyvale_total_pop

Race   Total People
0                    White  186864.581317
1                    Black   25699.914900
2  American Indian/Alaskan    1527.030000
3                    Asian  140181.354000
4          Native Hawaiian     610.812000
5              Two or More            NaN
6                 Hispanic  709693.662966

## Rate of Citation Per Race

In [31]:
asian = int(arrests[arrests['Race'].isin(['Asian','Asian Indian','Chinese','Filipino','Korean','Other Asian','Vietnamese'])]['Count'].sum())
pacific =  int(arrests[arrests['Race'].isin(['Guamanian','Native Hawaiian/Other Pacific','Pacific Islander','Hawiian','Samoan'])]['Count'].sum())

In [32]:
census_races_of_sunnyvale_citation = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(citations_race1920[citations_race1920['Race'] == 'White']['Count']), 
                             int(citations_race1920[citations_race1920['Race'] == 'Black']['Count']), 
                             int(citations_race1920[citations_race1920['Race'] == 'American Indian']['Count']), 
                             asian,
                             pacific,
                             0, 
                             int(citations_race1920[citations_race1920['Race'] == 'Hispanic']['Count'])]})
census_races_of_sunnyvale_citation

Race  Number of Residents
0                    White                 1590
1                    Black                  267
2  American Indian/Alaskan                   14
3                    Asian                  416
4          Native Hawaiian                   23
5              Two or More                    0
6                 Hispanic                 1787

In [33]:
rate = census_races_of_sunnyvale_citation.merge(sunnyvale_total_pop, on = 'Race')
rate['Rate'] = rate['Number of Residents'] / rate['Total People'] 
rate['Citations Relative To White Citations'] = rate['Rate'] / float(rate[rate['Race'] == 'White']['Rate'])
rate = rate[['Race', 'Rate', 'Citations Relative To White Citations']]
rate

Race      Rate  Citations Relative To White Citations
0                    White  0.008509                               1.000000
1                    Black  0.010389                               1.220983
2  American Indian/Alaskan  0.009168                               1.077483
3                    Asian  0.002968                               0.348765
4          Native Hawaiian  0.037655                               4.425376
5              Two or More       NaN                                    NaN
6                 Hispanic  0.002518                               0.295926

## Rate of Arrest per Race

In [34]:
asian = int(arrests[arrests['Race'].isin(['Asian','Asian Indian','Chinese','Filipino','Korean','Other Asian','Vietnamese'])]['Count'].sum())
pacific =  int(arrests[arrests['Race'].isin(['Guamanian','Native Hawaiian/Other Pacific','Pacific Islander','Hawiian','Samoan'])]['Count'].sum())

In [35]:
census_races_of_sunnyvale_arrests = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(arrests[arrests['Race'] == 'White']['Count']), 
                             int(arrests[arrests['Race'] == 'Black']['Count']), 
                             int(arrests[arrests['Race'] == 'American Indian/Alaskan']['Count']), 
                             asian,
                             pacific, 
                             0, 
                             int(arrests[arrests['Race'] == 'Hispanic/Mexican/Latin']['Count'])]})
census_races_of_sunnyvale_arrests

Race  Number of Residents
0                    White                 1144
1                    Black                  506
2  American Indian/Alaskan                    2
3                    Asian                  416
4          Native Hawaiian                   23
5              Two or More                    0
6                 Hispanic                 1803

In [36]:
rate = census_races_of_sunnyvale_arrests.merge(sunnyvale_total_pop, on = 'Race')
rate['Rate'] = rate['Number of Residents'] / rate['Total People'] 
rate['Arrests Relative To White Arrests'] = rate['Rate'] / float(rate[rate['Race'] == 'White']['Rate'])
rate = rate[['Race', 'Rate', 'Arrests Relative To White Arrests']]
rate

Race      Rate  Arrests Relative To White Arrests
0                    White  0.006122                           1.000000
1                    Black  0.019689                           3.216028
2  American Indian/Alaskan  0.001310                           0.213936
3                    Asian  0.002968                           0.484735
4          Native Hawaiian  0.037655                           6.150653
5              Two or More       NaN                                NaN
6                 Hispanic  0.002541                           0.414979

In [37]:
float(rate[rate['Race'] == 'White']['Rate'])

0.006122080449569328

# Conclusions

The conclusion this seems to draw is that Black people are 1.22 times as likely to be cited as White people, and that and that Black people are 3.22 times as likely to be arrested as White people.

That being said, it is unlikely this is a completely accurate statistical representation. There are a multitude of factors that contribute to the likely inaccuracy. 

1) Being unable to determine residents vs. non residents. This makes it impossible to correctly draw conclusions about the exact rate of arrests/citations, as explained at the top of this section.
2) The large number of people who were categorized as "Other" or "Unknown" or not recorded at all. 
3) The categorization of people as "Two or More" races in the census while also likely double counting them in other races. 

There are likely many other factors I have forgotten. Please feel free to reach out to me with other thoughts on how to improve this. 

The purpose of this is not at all to draw any decisive conclusions. Instead, this is merely intended to show how much more data we need to collect before any definitive conclusions can be drawn, as well as call for a general greater accountability of all of our police departments.

# Estimate Racial Breakdown by Radar Speed Citations

Based on the idea that speed citations should be relatively free of bias, I found all citations that have violation codes for a speed law. That being said, there is the possibility of bias still, and this idea would need to be checked by an expert.

In [38]:
citations1920

Report Number       Date        Violation Code          Race   Sex
1          T503714   1/1/2019                 24601         Black  Male
2          T506100   1/1/2019        267085A 12500A      Hispanic  Male
3          E219189   1/2/2019         23123A 16028A       Unknown     M
4          E219190   1/2/2019                21453A       Unknown     F
5          E219191   1/2/2019                21453A       Unknown     M
...            ...        ...                   ...           ...   ...
1339       E234645  6/28/2020                21453A  Not Recorded     M
1340       E234646  6/28/2020  12500A 4000A1 16028A      Hispanic     M
1341       E234647  6/28/2020                21453A         White     M
1342       E234648  6/29/2020          22350 12500A      Hispanic     M
1343       E234479  6/30/2020                27151A   Other Asian     M

[10863 rows x 5 columns]

In [39]:
speed_citation_numbers = [str(i) for i in range(22348, 22413)]

In [40]:
speed_citations = citations1920.loc[(citations1920['Violation Code'].str.contains('|'.join(speed_citation_numbers), case=False))]
speed_citations

Report Number       Date       Violation Code          Race   Sex
6          E219192   1/2/2019                22350       Unknown     M
8          E219194   1/2/2019                22350       Unknown     M
28         E219262   1/3/2019               22349A         Black     M
29         E219263   1/3/2019               22349A      Hispanic     F
31         E219265   1/3/2019               22349A      Hispanic     F
...            ...        ...                  ...           ...   ...
1276       T511197  5/29/2020  27156B 26101 22349A      Hispanic  Male
1295       E232805  6/12/2020          22350 21750  Not Recorded     M
1296       E234641  6/12/2020                22350  Not Recorded     M
1319       T508954  6/22/2020                22350      Hispanic  Male
1342       E234648  6/29/2020         22350 12500A      Hispanic     M

[891 rows x 5 columns]

In [41]:
race_speed_citations = speed_citations.groupby(['Race'])['Race'].count()
race_speed_citations = pd.DataFrame(race_speed_citations)
race_speed_citations.columns = ['Count']
race_speed_citations.reset_index(inplace = True)
race_speed_citations

Race  Count
0  American Indian      1
1     Asian Indian     96
2            Black     16
3         Hispanic    145
4     Not Recorded    185
5            Other     72
6      Other Asian    157
7          Unknown     90
8            White    129

In [42]:
asian = int(race_speed_citations[race_speed_citations['Race'].isin(['Asian','Asian Indian','Chinese','Filipino','Korean','Other Asian','Vietnamese'])]['Count'].sum())
pacific =  int(race_speed_citations[race_speed_citations['Race'].isin(['Guamanian','Native Hawaiian/Other Pacific','Pacific Islander','Hawiian','Samoan'])]['Count'].sum())

In [43]:
census_races_of_race_proxy = pd.DataFrame(
    {'Race': ['White', 'Black', 'American Indian/Alaskan', 'Asian', 'Native Hawaiian', 'Two or More', 'Hispanic'],
     'Number of Residents': [int(race_speed_citations[race_speed_citations['Race'] == 'White']['Count']), 
                             int(race_speed_citations[race_speed_citations['Race'] == 'Black']['Count']), 
                             int(race_speed_citations[race_speed_citations['Race'] == 'American Indian']['Count']), 
                             asian,
                             pacific,
                             0,
                             int(race_speed_citations[race_speed_citations['Race'] == 'Hispanic']['Count'])]})
census_races_of_race_proxy

Race  Number of Residents
0                    White                  129
1                    Black                   16
2  American Indian/Alaskan                    1
3                    Asian                  253
4          Native Hawaiian                    0
5              Two or More                    0
6                 Hispanic                  145

When comparing the racial breakdown of speed citations by percentage to the Sunnyvale census breakdown by percentage, it ended up being pretty similar.

In [44]:
total_count = census_races_of_race_proxy['Number of Residents'].sum()

In [45]:
census_races_of_race_proxy['Rate'] = census_races_of_race_proxy['Number of Residents'] / total_count
census_races_of_race_proxy = census_races_of_race_proxy[['Race', 'Number of Residents', 'Rate']]
census_races_of_race_proxy

Race  Number of Residents      Rate
0                    White                  129  0.237132
1                    Black                   16  0.029412
2  American Indian/Alaskan                    1  0.001838
3                    Asian                  253  0.465074
4          Native Hawaiian                    0  0.000000
5              Two or More                    0  0.000000
6                 Hispanic                  145  0.266544

In [46]:
total_citations = census_races_of_sunnyvale_citation['Number of Residents'].sum()
census_races_of_sunnyvale_citation['Percentage'] = census_races_of_sunnyvale_citation['Number of Residents']/total_citations

In [47]:
rate = census_races_of_sunnyvale_citation.merge(census_races_of_race_proxy, on = 'Race')
rate['Rate'] = rate['Percentage'] / rate['Rate'] 
rate['Citations Rate Relative To White Rate Citations'] = rate['Rate'] / float(rate[rate['Race'] == 'White']['Rate'])
rate = rate[['Race', 'Citations Rate Relative To White Rate Citations']]
rate

Race      Rate  Citations Relative To White Citations
0                    White  1.636592                               1.000000
1                    Black  2.215768                               1.353892
2  American Indian/Alaskan  1.858921                               1.135849
3                    Asian  0.218326                               0.133403
4          Native Hawaiian       inf                                    inf
5              Two or More       NaN                                    NaN
6                 Hispanic  1.636400                               0.999883

In [48]:
total_arrests = census_races_of_sunnyvale_arrests['Number of Residents'].sum()
census_races_of_sunnyvale_arrests['Percentage'] = census_races_of_sunnyvale_arrests['Number of Residents']/total_citations

In [49]:
rate = census_races_of_sunnyvale_arrests.merge(census_races_of_race_proxy, on = 'Race')
rate['Rate'] = rate['Percentage'] / rate['Rate'] 
rate['Arrests Relative To White Arrests'] = rate['Rate'] / float(rate[rate['Race'] == 'White']['Rate'])
rate = rate[['Race', 'Arrest Rate Relative To White Arrest Rate']]
rate

Race      Rate  Arrests Relative To White Arrests
0                    White  1.177523                           1.000000
1                    Black  4.199170                           3.566106
2  American Indian/Alaskan  0.265560                           0.225524
3                    Asian  0.218326                           0.185411
4          Native Hawaiian       inf                                inf
5              Two or More       NaN                                NaN
6                 Hispanic  1.651052                           1.402140